File: codes/0103-Binary-Tree-Zigzag-Level-Order-Traversal/cpp/main3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if (!root) return {};
        vector<vector<int>> res;
        queue<TreeNode*> q;
        q.push(root);
        int height = 1;
        while (!q.empty()) {
            int n = q.size();
            vector<int> sub;
            for (int i = 0; i < n; ++i) {
                TreeNode* t = q.front(); q.pop();
                if (height % 2 == 1) {
                    sub.push_back(t->val);
                } else {
                    sub.insert(sub.begin(), t->val);
                }
                if (t->left) q.push(t->left);
                if (t->right) q.push(t->right);
            }
            ++height;
            res.push_back(sub);
        }
        return res;        
    }
};

File: codes/0103-Binary-Tree-Zigzag-Level-Order-Traversal/Java/main3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
    	List<List<Integer>> res = new ArrayList<>();
    	if (root == null) return res;
    	Deque<TreeNode> q = new LinkedList<>();
    	q.addLast(root);
    	int height = 1;
    	while (q.size() > 0) {
    		List<Integer> sub = new ArrayList<>();
    		int n = q.size();
    		while (n-- > 0) {
    			TreeNode t = q.removeFirst();
    			if (height % 2 == 1) {
    				sub.add(t.val);
    			}
    			else {
    				sub.add(0, t.val);
    			}

    			if (t.left != null) q.addLast(t.left);
    			if (t.right != null) q.addLast(t.right);
    		}
    		++height;
    		res.add(sub);
    	}
    	return res;
    }
}